### PR TITLE
operator: Fix storing authentication credentials in the Loki ConfigMap

### PR DESCRIPTION
**Pull request by periklis**
**What this PR does / why we need it**:
The following PR replaces all sensitive authentication information from the Loki configuration file by reading values from environment variables. Latter are projected from the object storage secret via `EnvVar.ValueFrom` source declaration. 

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:
- The environment variables are projected into the ConfigMap cause of Loki's inability to leave the underlying SDKs reading their values instead of passing them via it's config.
- For testing on Azure you can consider using the scripts from this PR: 
  - #11340

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [x] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: operator/internal/handlers/internal/storage/secrets.go [hunk 1]
```diff
@@ -129,12 +125,12 @@ func extractS3SSEConfig(d map[string][]byte) (storage.S3SSEConfig, error) {
 		kmsKeyId, kmsEncryptionCtx string
 	)
 
-	switch sseType = storage.S3SSEType(d["sse_type"]); sseType {
+	switch sseType = storage.S3SSEType(d[storage.KeyAWSSSEType]); sseType {
 	case storage.SSEKMSType:
-		kmsEncryptionCtx = string(d["sse_kms_encryption_context"])
-		kmsKeyId = string(d["sse_kms_key_id"])
+		kmsEncryptionCtx = string(d[storage.KeyAWSSSEKMSEncryptionContext])
+		kmsKeyId = string(d[storage.KeyAWSSSEKMSKeyID])
```
Comment (xperimental) [inline operator/internal/handlers/internal/storage/secrets.go:131]: I do know that these are all valid acronyms written together, but this _is_ an ugly constant name.

File: operator/internal/handlers/internal/storage/secrets.go [hunk 2]
```diff
@@ -154,57 +150,55 @@ func extractS3SSEConfig(d map[string][]byte) (storage.S3SSEConfig, error) {
 
 func extractSwiftConfigSecret(s *corev1.Secret) (*storage.SwiftStorageConfig, error) {
 	// Extract and validate mandatory fields
-	url := s.Data["auth_url"]
+	url := s.Data[storage.KeyOSSwiftAuthURL]
```
Comment (xperimental) [inline operator/internal/handlers/internal/storage/secrets.go:153]: Does it make sense to reduce clutter and remove "OS" from all the Swift constants?

File: operator/internal/manifests/storage/configure.go [hunk 3]
```diff
@@ -43,26 +34,27 @@ func ConfigureDeployment(d *appsv1.Deployment, opts Options) error {
...
 }
 
-// ConfigureDeployment merges a GCS Object Storage volume into the deployment spec.
-// With this, the deployment will expose an environment variable for Google authentication.
-func configureDeployment(d *appsv1.Deployment, secretName string) error {
-	p := ensureCredentialsForGCS(&d.Spec.Template.Spec, secretName)
+// ConfigureDeployment merges the object storage secret volume into the deployment spec.
+// With this, the deployment will expose credentials specific environment variables.
+func configureDeployment(d *appsv1.Deployment, secretName string, t lokiv1.ObjectStorageSecretType) error {
+	p := ensureObjectStoreCredentials(&d.Spec.Template.Spec, secretName, t)
```
Comment (xperimental) [inline operator/internal/manifests/storage/configure.go:57]: Use a longer parameter name?

```suggestion
func configureDeployment(d *appsv1.Deployment, secretName string, storeType lokiv1.ObjectStorageSecretType) error {
	p := ensureObjectStoreCredentials(&d.Spec.Template.Spec, secretName, storeType)
```

File: operator/internal/manifests/storage/configure.go [hunk 4]
```diff
@@ -105,7 +105,7 @@ func configureStatefulSetCA(s *appsv1.StatefulSet, tls *TLSConfig) error {
 	return nil
 }
 
-func ensureCredentialsForGCS(p *corev1.PodSpec, secretName string) corev1.PodSpec {
+func ensureObjectStoreCredentials(p *corev1.PodSpec, secretName string, t lokiv1.ObjectStorageSecretType) corev1.PodSpec {
```
Comment (xperimental) [inline operator/internal/manifests/storage/configure.go:108]:
```suggestion
func ensureObjectStoreCredentials(p *corev1.PodSpec, secretName string, storeType lokiv1.ObjectStorageSecretType) corev1.PodSpec {
```

File: operator/internal/manifests/storage/configure.go [hunk 5]
```diff
@@ -124,10 +124,118 @@ func ensureCredentialsForGCS(p *corev1.PodSpec, secretName string) corev1.PodSpe
 		MountPath: secretDirectory,
 	})
 
-	container.Env = append(container.Env, corev1.EnvVar{
-		Name:  EnvGoogleApplicationCredentials,
-		Value: path.Join(secretDirectory, GCSFileName),
-	})
+	var objStoreEnvVar []corev1.EnvVar
+	switch t {
```
Comment (xperimental) [inline operator/internal/manifests/storage/configure.go:128]:
```suggestion
	switch storeType {
```

**Review by xperimental — COMMENTED** [review]
Looks good overall. Only tested it with S3 so far. Found a few typos, see comments.

**Review by xperimental — COMMENTED** [review]
While testing today something occurred to me, that I did not think of yesterday: This change makes the configuration file static with regards to some of the fields of the secret (the ones that are passed using environment variables now).
We currently pass an annotation containing the hash of the configuration file to all the deployments / statefulsets that we create, so that they automatically re-create their pods when the configuration changes. Because some fields of the secret now do not correspond to a change in the configuration file it can happen that a change to the secret does not trigger an update of the deployments/statefulsets anymore causing old configuration to persist in the Loki instances.
I guess a simple fix for this would be to take the contents of the secret into account when creating the "config-hash" as well. Adding a new annotation is also an alternative.

File: operator/internal/manifests/internal/config/loki-config.yaml [hunk 6]
```diff
@@ -37,7 +37,7 @@ common:
         kms_key_id: {{ .KMSKeyID }}
         {{- with .KMSEncryptionContext }}
         kms_encryption_context: |
-          {{ . }}
+          ${AWS_SSE_KMS_ENCRYPTION_CONTEXT}
```
[the remark this answers is not in the document]
Comment (periklis) [reply operator/internal/manifests/internal/config/loki-config.yaml:40]: :thinking:

File: operator/internal/handlers/internal/storage/secrets.go [hunk 7]
```diff
@@ -37,86 +43,118 @@ func ExtractSecret(s *corev1.Secret, secretType lokiv1.ObjectStorageSecretType)
...
 	endpointSuffix := s.Data[storage.KeyAzureStorageEndpointSuffix]
 
+	contents := bytes.Join([][]byte{env, container, name, key, endpointSuffix}, hashSeparator)
+
+	h := sha1.New()
+	_, err := h.Write([]byte(contents))
+	if err != nil {
+		return nil, "", kverrors.Wrap(err, "failed hashing azure secret contents")
+	}
+	sha1C := fmt.Sprintf("%x", h.Sum(nil))
```
Comment (xperimental) [inline operator/internal/handlers/internal/storage/secrets.go:75]: I think we can get away with not doing the hash calculation specially for each secret type but just hash all the data contents of the secret ([code suggestion](https://github.com/xperimental/loki/commit/9173a7d52ab9ce2709d14ca59c23be805afc4dbf)).

File: operator/internal/manifests/compactor_test.go [hunk 8]
```diff
@@ -48,10 +49,32 @@ func TestNewCompactorStatefulSet_HasTemplateConfigHashAnnotation(t *testing.T) {
...
+				Compactor: &lokiv1.LokiComponentSpec{
+					Replicas: 1,
+				},
+			},
+		},
+	})
+
+	annotations := ss.Spec.Template.Annotations
+	require.Contains(t, annotations, AnnotationLokiObjectStoreHash)
+	require.Equal(t, annotations[AnnotationLokiObjectStoreHash], "deadbeef")
```
Comment (xperimental) [inline operator/internal/manifests/compactor_test.go:77]: Should we just add this `require.Equal` to the other annotations test?

**Review by xperimental — COMMENTED** [review]
Pod updates seem to work fine now :+1: 

One refactoring idea in the comments, looks fine for me otherwise.

File: operator/internal/handlers/internal/storage/secrets.go (hunk 7 above)
Comment (periklis) [reply operator/internal/handlers/internal/storage/secrets.go:75]: Will apply that suggestion! Looks really nice

**Pull request merged**